src/math/exp_fcn_hifi.c

```diff
@@ -357,20 +312,27 @@ int32_t sofm_exp_fixed(int32_t x)
 	xs = x;
 	while (xs >= SOFM_EXP_TWO_Q27 || xs <= SOFM_EXP_MINUS_TWO_Q27) {
 		xs >>= 1;
-		n++;
+		n <<= 1;
 	}
```

Comment:
can we not just count the number of bits to shift and shift by N directly? Maybe with a branch like
```
if (xs >= SOFM_EXP_TWO_Q27)
    xs >>= ffs(xs) - ffs(SOFM_EXP_TWO_Q27);
else if (xs <= SOFM_EXP_MINUS_TWO_Q27)
    xs >>= ...;
```
? The actual shift above is likely wrong, but that would be the idea

Reply:
Also, it's max. 3 loops for worst case large input number. So, the overhead from counting bits other way needs to be small.

Reply:
@singalsu on Xtensa `__builtin_ffs()` uses the NSAU instruction to count those bits

Reply:
Thanks, I didn't know, I'll try with it.

Reply:
@singalsu yep, we have definitions for all the `ffs*()` and `clz*()` variants in https://github.com/thesofproject/sof/blob/53f2d9692ae49f8d740dab2d6b3ff9f0a2c3a7cc/src/include/sof/common.h#L133-L139

Reply:
OK, I'm able to save another 0.4 MCPS with normalize HiFi instruction based input value scaling to replace the while loop.  Also even more saving seems to be possible with allowing range -4 to +4 instead of -2 to +2 for the small value exponent function. I'll update this PR tomorrow.